### ISSUE TITLE
Potential information leaks in SRNS_syn?

### ISSUE THREAD

**Issue by VicaYang** [dingjingtao/SRNS#6]
In [line 180 of SRNS_syn/main.py](https://github.com/dingjingtao/SRNS/blob/e00f4e8783eedc57abc3b34d3131151d2b42dd12/SRNS_syn/main.py#L180)
it checks that the negative samples should not be in the train set.
```{python}
while negitems_candidates[i,0] in train_set[user_batch[batch_idx][I]] : 
    negitems_candidates[i,0] = random.randint(0,num_item-1)
```
However, in [line 216 of SRNS_syn/main.py](https://github.com/dingjingtao/SRNS/blob/e00f4e8783eedc57abc3b34d3131151d2b42dd12/SRNS_syn/main.py#L216), it checks that it checks that the negative samples should not be in both the train set and the test set.
```{python}
while random_item in train_set[user_batch[batch_idx][i]] or random_item in test_set[user_batch[batch_idx][I]]:
    random_item = random.randint(0,num_item-1)
```
I have concerns that the information from the test set will leak during the training. Is there any special reason to do that?

**Comment by dingjingtao** [comment]
> In [line 180 of SRNS_syn/main.py](https://github.com/dingjingtao/SRNS/blob/e00f4e8783eedc57abc3b34d3131151d2b42dd12/SRNS_syn/main.py#L180) it checks that the negative samples should not be in the train set.
> 
> ```
> while negitems_candidates[i,0] in train_set[user_batch[batch_idx][I]] : 
>     negitems_candidates[i,0] = random.randint(0,num_item-1)
> ```
> 
> However, in [line 216 of SRNS_syn/main.py](https://github.com/dingjingtao/SRNS/blob/e00f4e8783eedc57abc3b34d3131151d2b42dd12/SRNS_syn/main.py#L216), it checks that it checks that the negative samples should not be in both the train set and the test set.
> 
> ```
> while random_item in train_set[user_batch[batch_idx][i]] or random_item in test_set[user_batch[batch_idx][I]]:
>     random_item = random.randint(0,num_item-1)
> ```
> 
> I have concerns that the information from the test set will leak during the training. Is there any special reason to do that?

Hi Weikai, thanks for your attention to our work. Actually, this code in SRNS_syn is related to our synthetic adding-noise experiments in the NeurIPS'20 paper. In the experiments, we aim to evaluate whether SRNS is robust to our injected false-negative (FN) noise. Therefore we intentionally add FNs in L213, and filter out positives, i.e., FNs, in L216.

**Comment by VicaYang** [comment]
Thanks for your quick reply. I am a little bit confused though. If `args.fn_num` is small, the memory in SRNS will be cleaner than uniform, is my understanding correct? Just want to double-check because it seems a little wizard to me lol.

BTW, I have two more concerns after analyzing the code.
1. How the `test_merge.pkl` is constructed? It seems that it is a random combination of the training set and test set. If so, I wonder why we evaluate the performance on this set instead of a pure test set.
2. I used to think that RECALL@1 <= RECALL@3 <= RECALL@5, but the results say no. I think it might be related to "LgtItem_final," which is the min of K and #TP. Though looks reasonable, it somehow looks more like precision than recall since #TP is usually larger in the synthetic experiment. I just want to check if it is reasonable.
3. Since `test_merge` have an overlap with the `train_set,` and the rank calculation is ignoring the item in `train_set`, I am not sure whether the calculation is still correct. For example, let both the train_set and `test_merge` of $u_0$ contain $i_0$ and $i_1$, and their scores are the largest two. Then the hr_res will be [[1,1,1],[1,1,1]]? and the RECALL@1 might be 2.

TBH, I am new to the recommender system so some of my questions may look stupid. Hope they do not bother you much.

**Comment by dingjingtao** [comment]
> Thanks for your quick reply. I am a little bit confused though. If `args.fn_num` is small, the memory in SRNS will be cleaner than uniform, is my understanding correct? Just want to double-check because it seems a little wizard to me lol.
> 
> BTW, I have two more concerns after analyzing the code.
> 
> 1. How the `test_merge.pkl` is constructed? It seems that it is a random combination of the training set and test set. If so, I wonder why we evaluate the performance on this set instead of a pure test set.
> 2. I used to think that RECALL@1 <= RECALL@3 <= RECALL@5, but the results say no. I think it might be related to "LgtItem_final," which is the min of K and #TP. Though looks reasonable, it somehow looks more like precision than recall since #TP is usually larger in the synthetic experiment. I just want to check if it is reasonable.
> 3. Since `test_merge` have an overlap with the `train_set,` and the rank calculation is ignoring the item in `train_set`, I am not sure whether the calculation is still correct. For example, let both the train_set and `test_merge` of u0 contain i0 and i1, and their scores are the largest two. Then the hr_res will be [[1,1,1],[1,1,1]]? and the RECALL@1 might be 2.
> 
> TBH, I am new to the recommender system so some of my questions may look stupid. Hope they do not bother you much.

For your question on FN number, please check Figure 2 in our paper. We tested different levels ($\sigma$) of noise.

1. / 3. We only save ground-truth interactions of the test set in `test_merge.pkl`. 'test_merge[uid]' represents the uid's interacted item list in the test set.
2. Your understanding is correct. When N in Recall@N is small, like 1, this metric is close to precision. Actually, we also report a large N (5/10) in Appendix C.4.2.

**Comment by VicaYang** [comment]
So what's the difference between `ml-100k` and `ml-100k_tuning`? It seems that test_merge is identical to test for `ml-100k`, but not for `ml-100k-tuning.`

```{python}
import pickle
import numpy as np
def load_pickle(fname):
    with open(fname, 'rb') as f:
        return pickle.load(f)

DIR = 'ml-100k_tuning'
train_data = load_pickle(f'{DIR}/train.pkl')
test_data = load_pickle(f'{DIR}/test.pkl')
test_merge = load_pickle(f'{DIR}/test_merge.pkl')

train_set, test_set = dict(), dict()
num_user = max(np.max(train_data[:,0]), np.max(train_data[:,0])) + 1
for i in range(num_user):
    train_set[i] = set()
    test_set[i] = set()
for u, i in train_data:
    train_set[u].add(i)
for u, i in test_data:
    test_set[u].add(i)

UID = 0
S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
print(len(S1), len(S2), len(S3)) # 103 60 79
print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 50 29

UID = 10
S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
print(len(S1), len(S2), len(S3)) # 56 30 37
print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 26 11
```

**Comment by QuanYuhan** [comment]
> So what's the difference between `ml-100k` and `ml-100k_tuning`? It seems that test_merge is identical to test for `ml-100k`, but not for `ml-100k-tuning.`
> 
> ```
> import pickle
> import numpy as np
> def load_pickle(fname):
>     with open(fname, 'rb') as f:
>         return pickle.load(f)
> 
> DIR = 'ml-100k_tuning'
> train_data = load_pickle(f'{DIR}/train.pkl')
> test_data = load_pickle(f'{DIR}/test.pkl')
> test_merge = load_pickle(f'{DIR}/test_merge.pkl')
> 
> train_set, test_set = dict(), dict()
> num_user = max(np.max(train_data[:,0]), np.max(train_data[:,0])) + 1
> for i in range(num_user):
>     train_set[i] = set()
>     test_set[i] = set()
> for u, i in train_data:
>     train_set[u].add(i)
> for u, i in test_data:
>     test_set[u].add(i)
> 
> UID = 0
> S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
> print(len(S1), len(S2), len(S3)) # 103 60 79
> print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 50 29
> 
> UID = 10
> S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
> print(len(S1), len(S2), len(S3)) # 56 30 37
> print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 26 11
> ```

We have confirmed that there is an error in the `test_merge.pkl` file in `ml-100k_tuning` and we have updated the file so that it meets the requirements we mentioned above

> 1. / 3. We only save ground-truth interactions of the test set in `test_merge.pkl`. 'test_merge[uid]' represents the uid's interacted item list in the test set.

**Comment by dingjingtao** [comment]
> So what's the difference between `ml-100k` and `ml-100k_tuning`? It seems that test_merge is identical to test for `ml-100k`, but not for `ml-100k-tuning.`
> 
> ```
> import pickle
> import numpy as np
> def load_pickle(fname):
>     with open(fname, 'rb') as f:
>         return pickle.load(f)
> 
> DIR = 'ml-100k_tuning'
> train_data = load_pickle(f'{DIR}/train.pkl')
> test_data = load_pickle(f'{DIR}/test.pkl')
> test_merge = load_pickle(f'{DIR}/test_merge.pkl')
> 
> train_set, test_set = dict(), dict()
> num_user = max(np.max(train_data[:,0]), np.max(train_data[:,0])) + 1
> for i in range(num_user):
>     train_set[i] = set()
>     test_set[i] = set()
> for u, i in train_data:
>     train_set[u].add(i)
> for u, i in test_data:
>     test_set[u].add(i)
> 
> UID = 0
> S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
> print(len(S1), len(S2), len(S3)) # 103 60 79
> print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 50 29
> 
> UID = 10
> S1, S2, S3 = train_set[UID], test_set[UID], set(test_merge[UID])
> print(len(S1), len(S2), len(S3)) # 56 30 37
> print(len(S1&S2), len(S1&S3), len(S2&S3)) # 0 26 11
> ```

Thanks for this careful inspection.  We provided 'ml-100k_tuning' for hyperparameter tuning. 'ml-100k' is the formal dataset.

**Comment by VicaYang** [comment]
Thanks for the patient replies! I have no more questions.
